direct/nn/conjgradnet/conjgrad.py

```diff
@@ -0,0 +1,323 @@
+# coding=utf-8
+# Copyright (c) DIRECT Contributors
+
+from enum import Enum
+from typing import Callable, List
+
+import torch
+from torch import nn
+
+from direct.data.transforms import (
+    complex_division,
+    complex_dot_product,
+    complex_multiplication,
+    expand_operator,
+    reduce_operator,
+)
+
+
+class CGUpdateType(str, Enum):
+    FR = "FR"
+    PRP = "PRP"
+    DY = "DY"
+    BAN = "BAN"
+
+
+class ConjGrad(nn.Module):
+    r"""Performs the Conjugate Gradient (CG) algorithm to approach a solution to:
+
+    .. math ::
+
+        \min_{x} f(x) = \min_{x} \fraq{1}{2} \big( ||\mathcal{A}(x) - y||_2^2 + \lambda ||x - z||_2^2 \big)
+
+    or equivalently solving the normal equation of the above:
+
+    .. math ::
+
+        \mathcal{B}(x): = \big(\mathcal{A} \circ \mathcal{A}^{*} + \lambda I\big) (x)
+        = \mathcal{A}^{*}(y) + \lambda z =: b.
+
+    Notes
+    -----
+    :class:`ConjGrad` has no trainable parameters. However, PyTorch ensures that gradients are computed.
+    """
+
+    def __init__(
+        self,
+        forward_operator: Callable,
+        backward_operator: Callable,
```

Comment:
I find Callable overly broad, perhaps we can define a `Operator` abstract class.

Reply:
This is how operators were defined since beggining of direct. Please create an issue or so make a new branch and fix it